Struct hover + workspace/reference output corrections

slimsag opened the pull request:
See individual commit messages for details.

keegancsmith commented on langserver/langserver_test.go:
```diff
@@ -83,7 +83,8 @@ func TestServer(t *testing.T) {
 			},
 			wantHover: map[string]string{
-			// "a.go:1:28": "(T).F string", // TODO(sqs): see golang/hover.go; this is the output we want
+				// "a.go:1:28": "(T).F string", // TODO(sqs): see golang/hover.go; this is the output we want
```
is this comment still relevant?

slimsag replied:
Yep, it is. This is still something we want to do, just not done via this PR.

keegancsmith commented on langserver/hover.go:
```diff
@@ -44,7 +44,15 @@ func (h *LangHandler) handleHover(ctx context.Context, conn JSONRPC2Conn, req *j
+			typ := obj.Type().Underlying()
+			if _, ok := typ.(*types.Struct); ok {
+				s = "type " + obj.Name() + " struct"
+			}
```
I wonder if you can just do
```
s = fmt.Sprintf("type %s %s", obj.Name(), obj.Type().Underlying())
```
since `obj.Type().Underlying()` is a `types.Type` interface which requires `String() string`.

slimsag replied:
I tried it just now, and the answer is no. The actual problem is that `obj.Type().Underlying().String()` is actually what returns the 'concrete type' i.e. `struct{...all fields...}` instead of just `struct` which reads a lot better.

```
--- FAIL: TestServer (0.04s)
    --- FAIL: TestServer/go_detailed (0.00s)
        --- FAIL: TestServer/go_detailed/hover-a.go:1:17 (0.00s)
        	langserver_test.go:560: got "type T struct{F string}", want "type T struct"
```

keegancsmith commented on langserver/internal/refs/refs.go:
```diff
@@ -255,6 +262,16 @@ func (c *Config) defInfo(pkg *types.Package, info *types.Info, rootFile *ast.Fil
+		return name.Name
+	}
+	split := strings.Split(importPath, "/")
+	return split[len(split)-1]
```
not sure if this is optimising too soon, but you can avoid extra allocations by using `strings.LastIndex`.

slimsag replied:
That def. would be optimizing too soon, and probably for something that doesn't matter, BUT that reads a lot cleaner IMO so done :)

keegancsmith commented on langserver/internal/refs/refs_test.go:
```diff
@@ -71,49 +71,49 @@ func TestParseFile(t *testing.T) {
-				&Ref{Def: Def{ImportPath: "net/http", Path: "Client"}, Position: pos("testdata/vars.go:6:14 (offset 74)")},
-				&Ref{Def: Def{ImportPath: "net/http", Path: "RoundTripper"}, Position: pos("testdata/vars.go:8:12 (offset 124)")},
-				&Ref{Def: Def{ImportPath: "net/http", Path: "Client Transport"}, Position: pos("testdata/vars.go:12:3 (offset 225)")},
-				&Ref{Def: Def{ImportPath: "net/http", Path: "Client"}, Position: pos("testdata/vars.go:11:12 (offset 194)")},
```
the updated test cases don't really imply why this is important (later commits might though). From what I can tell this can just be computed from `ImportPath`, and if not, we should add a test for that.

slimsag replied:
Good point. I will add a test case for it. It's because there can be both `package foo` and `package foo_test` in the same directory.

slimsag replied:
Also, you encouraging me to write a test for this also just made me uncover a bug that makes:

```
// packageName returns name.Name if name != nil. Otherwise, it returns the last
// element in importPath (e.g. "http" for "net/http")
func packageName(importPath string, name *ast.Ident) string {
	if name != nil {
		return name.Name
	}
	return importPath[strings.LastIndex(importPath, "/")+1:]
}
```

completely unneeded / incorrect :)

slimsag commented:
Addressed all feedback + added `go test -i ./...` to solve a (very confusing) CI problem 😄 (thanks @shurcooL for helping me solve it!)

Merging.
